### PR TITLE
[WIP][3/N]/dtensor] allow distribute_tensor to shard DTensor

### PR DESCRIPTION
**Pull request by XilunWu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130551
* #130760
* #130239
* #126697
* #130587

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/distributed/_tensor/api.py [hunk 1]
```diff
@@ -649,6 +650,69 @@ def distribute_tensor(
...
+        # then passed to another distribute_tensor() call along with another 1-d device
+        # mesh, and potentially so on.
+        input_parent_mesh = _mesh_resources.get_parent_mesh(device_mesh)
+        dtensor_parent_mesh = _mesh_resources.get_parent_mesh(tensor.device_mesh)
+        right_to_left_distribute = (
+            device_mesh.ndim == 1  # the device mesh arg must be a 1-d mesh
+            and len(placements) == 1  # must be True for 1-d mesh
+            and input_parent_mesh is not None
+            and input_parent_mesh == dtensor_parent_mesh  # distribute over the same mesh
+        )
```
Comment (fegin) [inline torch/distributed/_tensor/api.py:666]: Can we avoid using another indention and instead move the below return tensor and raise exception here? Then we don't need another indention.

File: torch/distributed/_tensor/api.py [hunk 2]
```diff
@@ -649,6 +650,69 @@ def distribute_tensor(
...
+
+        if right_to_left_distribute:
+            placement = placements[0]
+            if not placement.is_shard():
+                raise NotImplementedError(
+                    "Calling distribute_tensor() on DTensor objects require the "
+                    f"placement be a Shard() placement. Input args: tensor={tensor}, "
+                    f"device_mesh={device_mesh}, placements={placements}."
+                )
```
Comment (fegin) [inline torch/distributed/_tensor/api.py:675]: `replicate()` is also required as we need to support HSDP?

File: torch/distributed/_tensor/api.py [hunk 3]
```diff
@@ -649,6 +650,69 @@ def distribute_tensor(
...
         # 2. check if device mesh and placements are the same
+
+        # use case: right-to-left tensor distribution i.e. nested distribute_tensor()
+        # over 1-d meshes for composable parallelisms. This branch specifically allows
+        # for FSDP2 sharding where the tensor is first sharded on a 1-d device mesh and
+        # then passed to another distribute_tensor() call along with another 1-d device
+        # mesh, and potentially so on.
+        input_parent_mesh = _mesh_resources.get_parent_mesh(device_mesh)
+        dtensor_parent_mesh = _mesh_resources.get_parent_mesh(tensor.device_mesh)
+        right_to_left_distribute = (
```
Comment (fegin) [inline torch/distributed/_tensor/api.py:661]: We don't actually check if we are sharding from right to left. If the input mesh is actually in the right of the tensor's submesh, we should raise an exception.

**Review by wanchaol — COMMENTED** [review]
Overall I think this is valuable to add to `distribute_tensor`, but only after the strided shard placement get matured (i.e. all DTensor ops are working with this placement or we merge this placement back to Shard placemnet). 

Before that happens, I think we should focus on making FSDP2 + TP to use it first (i.e. add the integration to FSDP init shard param path)

**Pull request closed without merging**